sgl-kernel/csrc/gemm/per_token_quant_fp8.cu

```diff
@@ -1,67 +1,81 @@
-#include <ATen/cuda/CUDAContext.h>
-
+// clang-format off
+#include <tuple>
 #include <cmath>
-#include <cub/block/block_reduce.cuh>
+#include <ATen/cuda/CUDAContext.h>
+#include <cute/algorithm/functional.hpp>
+#include <cute/algorithm/gemm.hpp>
+#include <cute/arch/cluster_sm90.hpp>
+#include <cute/tensor.hpp>
 #include <flashinfer/vec_dtypes.cuh>
+// clang-format on
 
 #include "utils.h"
 
-template <typename T>
+using namespace cute;
+
+static constexpr int kWarpSize = 32;
+
+// ---------------------------------------------------------------------------
+// 1. SmallWarp kernel — warp‑local, no shared memory
+//    • One warp handles one token.
+//    • Eight tokens per 256‑thread CTA.
+// ---------------------------------------------------------------------------
+template <typename T, int kTokensPerCTA = 8, int kVecSize = 16>
 __global__ void per_token_quant_fp8_kernel(
     const T* __restrict__ input,
     FP8_TYPE* __restrict__ output_q,
     float* __restrict__ output_s,
     const int64_t hidden_dim,
     const int64_t num_tokens) {
-  const int token_idx = blockIdx.x;
-  if (token_idx >= num_tokens) return;
-
-  const int tid = threadIdx.x;
-  const int block_dim = blockDim.x;
-
-  const T* token_input = input + token_idx * hidden_dim;
-  FP8_TYPE* token_output = output_q + token_idx * hidden_dim;
-
-  float max_value = 0.0f;
-
-  // We want to store 128 bits of data at a time. 16 = 128 / 8 bits
-  // Load is already vectorized, so 16 elements work for T.
-  const uint32_t VEC_SIZE = 16;
-  using vec_t = flashinfer::vec_t<T, VEC_SIZE>;
-  const int32_t num_vec_elems = hidden_dim / VEC_SIZE;
-
-  // Find max using vectorized loads
-  for (int32_t i = tid; i < num_vec_elems; i += block_dim) {
+  const int warp_id = threadIdx.x / kWarpSize;        // 0‑7  (8 warps)
+  const int lane_id = threadIdx.x & (kWarpSize - 1);  // 0‑31
+  const int token = blockIdx.x * kTokensPerCTA + warp_id;
+  if (token >= num_tokens) return;
+
+  // Global tensors for this token
+  auto gmem_in = make_tensor(input + token * hidden_dim, make_shape(hidden_dim));
```

Comment:
Do not use cutlass in this element-wise kernel, you can refer https://github.com/sgl-project/sglang/blob/main/sgl-kernel/csrc/gemm/per_token_group_quant_8bit.cu#L50-L52 to locate data pointer.